crates/gitbutler-core/src/types/tagged_string.rs

```diff
@@ -0,0 +1,60 @@
+use std::{fmt, marker::PhantomData, ops::Deref};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// Tagged string is designed to clarify the purpose of strings when used as a return type
+pub struct TaggedString<T>(String, PhantomData<T>);
+
+impl<T> From<String> for TaggedString<T> {
+    fn from(value: String) -> Self {
+        TaggedString(value, PhantomData)
+    }
+}
+
+impl<T> From<&str> for TaggedString<T> {
+    fn from(value: &str) -> Self {
+        TaggedString(value.to_string(), PhantomData)
+    }
+}
+
+impl<T> Deref for TaggedString<T> {
+    type Target = String;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'de, T> Deserialize<'de> for TaggedString<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        String::deserialize(deserializer).map(Into::into)
+    }
+}
+
+impl<T> Serialize for TaggedString<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<T> fmt::Display for TaggedString<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl<T> fmt::Debug for TaggedString<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+pub struct _ReferenceName;
+/// The name of a reference ie. `refs/heads/master`
+pub type ReferenceName = TaggedString<_ReferenceName>;
```

Comment:
It certainly depends on where the reference name is coming from, but there is types in `git2` and `gix` which would assure the user that it is a valid reference name as well, and if it is partial (like `main`) or full (like `HEAD` or `refs/heads/main`.

Nonetheless, it's probably good to start with this type, but try to keep in mind that ideally, even more specific types are used. My take on it is that internally, only fully qualified reference names should be used, and with them, probably the `gix::refs::FullName` or `gix::refs::Reference`.

Reply:
I've returned either git2::Branch or git2::Reference from the functions inside virtual.rs because they're just easier to consume again; I've only used ReferenceName inside the controller.rs files as those values need to be serialisable.

We do have our internal Refname struct, but there is a major logical flaw in it and ideally it should be removed.

Reply:
`git2` types might have the disadvantage that sometimes they carry a reference to the `'repo` , which might make them unusable. `gix` has owned types for that, but maybe it's too early to introduce them even.

Regarding serializing these types, that's usually quite straightforward with one of these `core::serde::` modules, and `bstr` has serde-support as well.

If in doubt, it's certainly easier to keep it as is and there is merit in that. If the comment could then make clear what values are allowed or usually in there, that would help as well.